File: src/test/java/no/fintlabs/consumer/resource/ResourceAspectTest.java
package no.fintlabs.consumer.resource;

import no.fintlabs.consumer.exception.IdentificatorNotFoundException;
import no.fintlabs.consumer.exception.ResourceNotFoundException;
import no.fintlabs.consumer.exception.ResourceNotWriteableException;
import no.fintlabs.consumer.resource.aspect.IdentifierAspect;
import no.fintlabs.consumer.resource.aspect.ResourceAspect;
import no.fintlabs.consumer.resource.aspect.WriteableAspect;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.kafka.core.KafkaAdmin;
import org.springframework.kafka.core.KafkaTemplate;

import static org.junit.jupiter.api.Assertions.assertThrows;

@SpringBootTest
public class ResourceAspectTest {

    @Autowired
    private IdentifierAspect identifierAspect;

    @Autowired
    private ResourceAspect resourceAspect;

    @Autowired
    private WriteableAspect writeableAspect;

    // Mocking the kafka behaviour

    @MockBean
    private KafkaTemplate<String, String> kafkaTemplate;

    @MockBean
    private KafkaAdmin kafkaAdmin;

    private final String validResource = "elevfravar";
    private final String validWriteableResource = "eksamensgruppe";

    @Test
    void testResourceSuccess() {
        resourceAspect.checkResource(validResource);
    }

    @Test
    void testResourceFailure_WhenResourceDoesNotExistInComponent() {
        assertThrows(ResourceNotFoundException.class, () -> resourceAspect.checkResource("asdfasdfasdf"));
    }

    @Test
    void testIdentifierSuccess() {
        identifierAspect.checkIdField(validResource, "systemid");
    }

    @Test
    void testIdentifierFailure_WhenIdFieldDoesNotMatchResource() {
        assertThrows(IdentificatorNotFoundException.class, () -> identifierAspect.checkIdField(validResource, "fodselsnummer"));
    }

    @Test
    void testWriteableAspectSuccess() {
        writeableAspect.checkWriteable(validWriteableResource);
    }

    @Test
    void testWriteableAspectFailure_WhenResourceIsNotWriteable() {
        assertThrows(ResourceNotWriteableException.class, () -> writeableAspect.checkWriteable(validResource));
    }

}

File: src/test/java/no/fintlabs/consumer/resource/ResourceAspectTest.kt
package no.fintlabs.consumer.resource

import io.mockk.every
import io.mockk.mockk
import no.fintlabs.consumer.exception.resource.IdentificatorNotFoundException
import no.fintlabs.consumer.exception.resource.ResourceNotFoundException
import no.fintlabs.consumer.exception.resource.ResourceNotWriteableException
import no.fintlabs.consumer.resource.aspect.IdentifierAspect
import no.fintlabs.consumer.resource.aspect.ResourceAspect
import no.fintlabs.consumer.resource.aspect.WriteableAspect
import no.fintlabs.consumer.resource.context.ResourceContext
import no.novari.metamodel.MetamodelService
import no.novari.metamodel.model.Resource
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows

class ResourceAspectTest {
    private val metamodelService = mockk<MetamodelService>()
    private val resourceContext = mockk<ResourceContext>()
    private val resourceAspect = ResourceAspect(metamodelService)
    private val identifierAspect = IdentifierAspect(resourceContext)
    private val writeableAspect = WriteableAspect(resourceContext)

    private val domain = "utdanning"
    private val pkg = "elev"
    private val resource = "elev"
    private val key = "utdanning_elev_elev"

    @Test
    fun `checkResource passes for a valid triple`() {
        every { metamodelService.getResource(domain, pkg, resource) } returns mockk<Resource>()
        assertDoesNotThrow { resourceAspect.checkResource(domain, pkg, resource) }
    }

    @Test
    fun `checkResource throws for an invalid triple`() {
        every { metamodelService.getResource(any(), any(), any()) } returns null
        assertThrows<ResourceNotFoundException> { resourceAspect.checkResource(domain, pkg, "nope") }
    }

    @Test
    fun `checkIdField passes when the id field exists`() {
        every { resourceContext.resourceHasIdField(key, "systemid") } returns true
        assertDoesNotThrow { identifierAspect.checkIdField(domain, pkg, resource, "systemid") }
    }

    @Test
    fun `checkIdField throws when the id field does not exist`() {
        every { resourceContext.resourceHasIdField(key, "fodselsnummer") } returns false
        assertThrows<IdentificatorNotFoundException> {
            identifierAspect.checkIdField(domain, pkg, resource, "fodselsnummer")
        }
    }

    @Test
    fun `checkWriteable passes for a writeable resource`() {
        every { resourceContext.resourceIsWriteable(key) } returns true
        assertDoesNotThrow { writeableAspect.checkWriteable(domain, pkg, resource) }
    }

    @Test
    fun `checkWriteable throws for a non-writeable resource`() {
        every { resourceContext.resourceIsWriteable(key) } returns false
        assertThrows<ResourceNotWriteableException> { writeableAspect.checkWriteable(domain, pkg, resource) }
    }
}
